### ISSUE TITLE
Support "extensions" property on received errors in next

### ISSUE THREAD

**Issue by blorenz** [FormidableLabs/urql#88]
urql is discarding the "errors" if a "data" is being returned.  My GraphQL backend is returning `{"errors":[{"message":"Please, enter valid credentials","locations":[{"line":2,"column":3}],"path":["tokenAuth"]}],"data":{"tokenAuth":null}}` when attempting to get a JWT while logging in with invalid credentials.  It appears that in this is in http-exchange.ts.https://github.com/FormidableLabs/urql/blob/5e113c45f730223cc2d11125aa968ebc4eb260fd/src/modules/http-exchange.ts#L66-L76 .  

It seems this issue of both data and errors was faced before in https://github.com/graphql-go/graphql/issues/103

**Comment by kitten** [comment]
This will be fixed in urql v0.3 ✨ You can check its branch/PR out here: https://github.com/FormidableLabs/urql/pull/72

And you can try it already. It's released as `v0.3.0-next1` under the `true` tag, so `yarn add urql@true` should do it (Sorry about the weird tag)

The version currently does not have any breaking changes, so everything else should work exactly the same :+1:

**Comment by jwright** [comment]
I ran into a few issues trying to use this branch in some production code. I needed to also have `errors` returned alongside `data` nodes. I wanted to report those here and I am open to submit a PR if you feel these would be useful.

1. Returning the errors only returns an array of messages. We create errors in our resolvers that basically correspond to validation errors on the object being mutated. These errors contain some extra data in the `extensions` key on the object. (mentioned in [forthcoming GraphQL spec](http://facebook.github.io/graphql/draft/#example-fce18)). This means that our front-end cannot render the errors appropriately. Is there a way we can allow for additional fields on the [`Error interface`](https://github.com/FormidableLabs/urql/blob/v0.3.0-next1/src/interfaces/error.ts).

2. I wrapped the `executeQuery` function to check for the existence of errors in the response and `reject` the promise instead of calling the `resolve` callback. Is this something the `executeQuery` function should do by default? I have found it easier to handle and use less diverging of code.

**Comment by kitten** [comment]
@jwright We could add support for more fields via `extensions`. Let's track this using this issue :+1:

Regarding 2), no, this is not something we'll do by default. There's a strict logic to when we can consider a request as "failed", but like in Apollo a successful query can still imply an error on the backend, so we have to consider more of our queries as succeeded rather than failed. This is just a conceptual issue as we don't actually call these results "successful" but instead "loaded".

In summary of how it works in 0.3 now that we support errors properly: https://github.com/FormidableLabs/urql/blob/next/src/modules/http-exchange.ts#L66-L76

- If we have some data but also some errors, it resolves normally
- If we only have errors we emit them, wrapped in a CombinedError
- Otherwise we throw a "fatal" error

**Comment by kitten** [comment]
@jwright Addressed in PR #95 against `next`. Mind taking a quick look? :)
